test/Microsoft.AspNet.Mvc.Core.Test/Formatters/OutputFormatterTests.cs

```diff
@@ -102,6 +103,62 @@ public static IEnumerable<object[]> SelectResponseCharacterEncodingData
         }
 
         [Fact]
+        public void WriteResponse_GetMediaTypeWithCharsetReturnsMediaTypeFromCache_IfEncodingIsUtf8()
+        {
+            // Arrange
+            var formatter = new TestOutputFormatter();
+
+            var formatterContext = new OutputFormatterWriteContext(
+                new DefaultHttpContext(),
+                new TestHttpResponseStreamWriterFactory().CreateWriter,
+                objectType: null,
+                @object: null)
+            {
+                ContentType = new StringSegment("application/json"),
+            };
+
+            formatter.SupportedMediaTypes.Add("application/json");
+            formatter.SupportedEncodings.Add(Encoding.UTF8);
+
+            // Act
+            formatter.WriteAsync(formatterContext);
+            var firstContentType = formatterContext.ContentType;
+
+            formatterContext.ContentType = new StringSegment("application/json");
+
+            formatter.WriteAsync(formatterContext);
+            var secondContentType = formatterContext.ContentType;
+
+            // Assert
+            Assert.Same(firstContentType.Buffer, secondContentType.Buffer);
+        }
+
+        [Fact]
+        public void WriteResponse_GetMediaTypeWithCharsetReplacesCharset_IfDifferentThanEncoding()
+        {
+            // Arrange
+            var formatter = new TestOutputFormatter();
+
+            var formatterContext = new OutputFormatterWriteContext(
+                new DefaultHttpContext(),
+                new TestHttpResponseStreamWriterFactory().CreateWriter,
+                objectType: null,
+                @object: null)
+            {
+                ContentType = new StringSegment("application/json; charset=utf-7"),
+            };
+
+            formatter.SupportedMediaTypes.Add("application/json");
+            formatter.SupportedEncodings.Add(Encoding.UTF8);
+
+            // Act
+            formatter.WriteAsync(formatterContext);
+
+            // Assert
+            Assert.Equal(new StringSegment("application/json; charset=utf-8"), formatterContext.ContentType);
```

Comment:
You still haven't added the case that I'm asking for.

ContentType: `"application/json; charset=utf-16"` - supported encoding: UTF-16

Check that the string you get out is the same as the one you put in. You took the time to handle this case specially, so add a test for it :+1: